test/integration/test_kubernetes_runner.py

```diff
@@ -308,20 +308,28 @@ def test_external_job_delete(self):
     def test_exit_code_127(self):
         inputs = {
             "failbool": True,
-            "sleepsecs": 20,
+            "sleepsecs": 5,
         }
         running_response = self.dataset_populator.run_tool_raw(
             "job_properties",
             inputs,
             self.history_id,
         )
-        time.sleep(10)
         # check that logs are also available in job logs
         app = self._app
-        sa_session = app.model.context.current
-        job = sa_session.query(app.model.Job).get(app.security.decode_id(running_response.json()["jobs"][0]["id"]))
+        job_id = app.security.decode_id(running_response.json()["jobs"][0]["id"])
+        sa_session = app.model.context
+        job = sa_session.query(app.model.Job).get(job_id)
+        self._wait_for_external_state(sa_session=sa_session, job=job, expected=app.model.Job.states.RUNNING)
+
         external_id = job.job_runner_external_id
-        output = unicodify(subprocess.check_output(["kubectl", "logs", "-l" f"job-name={external_id}"]))
+        log_cmd = ["kubectl", "logs", "-l", f"job-name={external_id}"]
+        p = subprocess.run(log_cmd, capture_output=True, text=True)
+        if p.returncode:
+            raise Exception(
+                f"Command '{' '.join(log_cmd)}' failed with exit code: {p.returncode}.\nstdout: {p.stdout}\nstderr: {p.stderr}"
```

Comment:
What about adding the following to `lib/galaxy/util/__init__.py`:

```python
import shlex

try:
    shlex_join = shlex.join
except AttributeError:
    # Python < 3.8
    shlex_join = lambda split_command: " ".join(map(shlex.quote, split_command))
```

and then:

```suggestion
                f"Command '{shlex_join(log_cmd)}' failed with exit code: {p.returncode}.\nstdout: {p.stdout}\nstderr: {p.stderr}"
```

We can reuse such `shlex_join()` also in `lib/galaxy/datatypes/text.py`:222 , `lib/galaxy/tool_util/deps/conda_util.py`:240, `lib/galaxy/tool_util/deps/mulled/mulled_build.py`:304 and `lib/galaxy/util/path/__init__.py`:368 .